src/error.rs

```diff
@@ -10,13 +11,17 @@ pub enum CommunicationError {
     #[error("Timeout")]
     Timeout,
     /// Implementation specific modbus error
+    #[cfg(feature = "tokio-modbus")]
     #[error("Modbus")]
-    Modbus(Box<dyn Error>),
```

Comment:
I'm not sure if I like this. I chose `Box<dyn Error>` so the error structs could be used with alternative modbus implementations. Do you really need to match on the error type or what is the rationale behind this?

With a `dyn Error` you can call [Error::downcast](https://doc.rust-lang.org/stable/std/error/trait.Error.html#method.downcast) and cast it to a concrete type.

I know it's not as nice as calling `match e { ... }` but for the rare cases where you actually need the error type its an option.

Alternatively `CommunicationError` (and the other error types) could add a generic error type for the underlying modbus error.

Reply:
My motivation is to fix the `(dyn StdError + 'static) cannot be shared between threads safely` errors I started getting after updating to `sunspec 0.5`. I don't think `Error::downcast` would help me with that, right...?

From my limited understanding of systems programming I think it is better to use non-boxed errors for performance. And personally I think it makes the code easier to read and understand.

*If* an implementation alternative to `tokio-modbus` were to be added then I think it would make sense to abstract it everything away more, with explicit error variants provided by `sunspec`.

Reply:
Ah, gotcha. All that is needed is a `+ Send`.

```diff
-    Modbus(Box<dyn Error>),
+    Modbus<Box<dyn Error + Send>),
```

Yes. I do like concrete error variants better, too. Though those errors should be rare and that extra allocation is not a huge price to pay.

I guess it would make the most sense to just move all errors into the `tokio_modbus` crate. There is little generic about them anyways.

Reply:
Yeah you're probably right about the extra allocation.

> I guess it would make the most sense to just move all errors into the tokio_modbus crate. There is little generic about them anyways.
Do you mean something like what I pushed just right now? (d83bbe3fa5bfe7c0c9c0d01fc47915be72c57642)